### PR TITLE
New Addon: `remaining-replies`

### PR DESCRIPTION
**Pull request by BarelySmooth**
### Changes
New addon that adds the number of remaining replies in a studio comment thread.
[Image demonstrating the addons function](https://scratch.mit.edu/studios/30131729/comments/#comments-181718673)
(ignore the typo in the screenshot)

~~The placement of the text should probably be reconsidered (it's a little confusing), but that was the best place I could think of.~~

### Tests
Tested on Edge (because firefox makes it really hard to test local extensions)

### CODE REVIEW

**Comment by mxmou** [comment]
> The placement of the text should probably be reconsidered (it's a little confusing), but that was the best place I could think of.

Could it be on the reply button of each comment in the thread? For example "reply (5 left)".

**Comment by WorldLanguages** [comment]
@BarelySmooth Note that mxmou liked your message (GitHub doesn't notify reactions)
But yeah it's repetitive, maybe it could only be shown after the thread has 10 replies? @mxmou What do you think?

**Comment by mxmou** [comment]
It could also be a tooltip on the reply button.

**Comment by BarelySmooth** [comment]
> But yeah it's repetitive, maybe it could only be shown after the thread has 10 replies?

Could this be an option within settings?

**Comment by Norbiros** [comment]
Please add developer comments to it, because translators can translate it as "left site" (in polish "lewo"), or as " 5 left' (in polish " zostało 5")...

**Comment by BarelySmooth** [comment]
> Please add developer comments to it, because translators can translate it as "left site" (in polish "lewo"), or as " 5 left' (in polish " zostało 5")...

How can I do that? (Adding it to the userscript probably wont work because the translation is done on Transifex)


https://github.com/ScratchAddons/ScratchAddons/blob/f1fddc7409085b7d63fd2990028fbacef7092b84/addons-l10n/en/remaining-replies.json#L3
I used the word "remaining" instead of "left" in the json to prevent confusion

**Comment by Norbiros** [comment]
Yesterday @apple502j added this feature... Maybe he know?

Okay. Thats better!

**Comment by BarelySmooth** [comment]
> But yeah it's repetitive, maybe it could only be shown after the thread has 10 replies?

@WorldLanguages I've added an option to the addon settings, is that fine?

screenshot
screenshot

**Review by DNin01 — APPROVED** [review]
Tried it, works well! The counter even updates after you reply 😃

Perhaps the addon could also gray out/dim the reply button when the comment thread has reached its limit.

**Comment by Endermaster3690** [comment]
> screenshot

is there a way to make it custom?

**Comment by BarelySmooth** [comment]
> Tried it, works well! The counter even updates after you reply 😃
>
> Perhaps the addon could also gray out/dim the reply button when the comment thread has reached its limit.

Thanks! 
Also, Greying out the reply button is possible, but that isn't the purpose of the addon (imo).

**Comment by lisa-wolfgang** [comment]
In my opinion, a better, less repetitive location for this is next to the Post/Cancel button area.

image

**Comment by BroJac5246** [comment]
> Another option would be to keep the number in the reply button, but only show the remaining replies text while it's being hovered.

I think that would be best.


> > Also, Greying out the reply button is possible, but that isn't the purpose of the addon (imo).
> 
> It seems like a good related feature to bundle in the same addon.

I agree.

File: addons-l10n/en/remaining-replies.json
Comment (Samq64) [inline addons-l10n/en/remaining-replies.json:3]: Merging the two messages into
```suggestion
  "remaining-replies/reply": "reply ({replies} left)"
```
should make things simpler in the JS file and possibly for translators.

**Review by Samq64 — COMMENTED** [review]
Could you add dynamic enable/disable support?

File: addons/remaining-replies/addon.json
Comment (DNin01) [reply addons/remaining-replies/addon.json:2]: This sounds more like replies to a studio, not replies to comments in studios. Maybe "remaining reply counter" and specify in the description that this is for studios? We don't need the name to 100% describe the addon.

File: addons/remaining-replies/addon.json
Comment (DNin01) [inline addons/remaining-replies/addon.json:3]: I have a few description ideas:
- Shows you how many more times a comment thread in a studio can receive a reply.
- Shows you how many more replies can be posted on each comment thread in a studio.
- Shows you each studio comment thread's number of remaining replies.
- Shows you the number of remaining replies for each comment thread in a studio.
- Shows you how many more times you can reply to each comment thread in a studio.

Which one do you think is the best?

Comment (BroJac5246) [reply addons/remaining-replies/addon.json:3]: > I have a few description ideas:
> 
> * Shows you how many more times a comment thread in a studio can receive a reply.
> * Shows you how many more replies can be posted on each comment thread in a studio.
> * Shows you each studio comment thread's number of remaining replies.
> * Shows you the number of remaining replies for each comment thread in a studio.
> * Shows you how many more times you can reply to each comment thread in a studio.
> 
> Which one do you think is the best?

I prefer "Shows you how many more replies can be posted on each comment thread in a studio."

**Comment by Samq64** [comment]
@BarelySmooth I committed the suggestions and made the addon dynamic.

File: addons/remaining-replies/userscript.js
Comment (Samq64) [inline addons/remaining-replies/userscript.js:23]: Would it be better to include a leading space in the message?

**Review by Samq64 — APPROVED** [review]

File: addons/remaining-replies/userscript.js
Comment (WorldLanguages) [reply addons/remaining-replies/userscript.js:23]: Probably not, as the leading space may get ignored by translators.

**Comment by WorldLanguages** [comment]
Comment sections on studios don't support RTL, right?

**Comment by mxmou** [comment]
> Comment sections on studios don't support RTL, right?

RTL support for studios was implemented but then reverted for some reason: https://github.com/scratchfoundation/scratch-www/commit/604238d

File: addons/remaining-replies/addon.json
Comment (WorldLanguages) [inline addons/remaining-replies/addon.json:31]:
```suggestion
  "versionAdded": "1.36.0",
```

File: addons/remaining-replies/addon.json
Comment (DNin01) [reply addons/remaining-replies/addon.json:3]: That's a good idea. I would personally think to say that in the description.

Comment (Samq64) [reply addons/remaining-replies/addon.json:3]:
```suggestion
  "description": "Shows you how many more replies can be posted on each comment thread in a studio before reaching the limit of 25.",
```

**Comment by WorldLanguages** [comment]
It's not clear to me why comments need to be observed through Redux and "waitForElement" at the same time, but it appears to work.

File: addons/remaining-replies/userscript.js
Comment (WorldLanguages) [inline addons/remaining-replies/userscript.js:20]: Shouldn't it check first if such span already exists in the comment?
This function "addRemainingReplyCount" can be called more than once on the same comment. It appears to be intentional.
This causes a bug

**Review by WorldLanguages — CHANGES_REQUESTED** [review]
Still buggy

**Comment by Samq64** [comment]
Should the reply threshold be customizable?

**Comment by Samq64** [comment]
> Shouldn't it check first if such span already exists in the comment? This function "addRemainingReplyCount" can be called more than once on the same comment. It appears to be intentional. This causes a bug image

I fixed that along with a few other edge cases. The only one left that I can think of is deleting a reply.

**Comment by WorldLanguages** [comment]
> Should the reply threshold be customizable?

@Samq64 Sorry, what do we mean by "reply threshold"?

**Review by TheColaber — APPROVED** [review]

**Comment by TheColaber** [comment]
@WorldLanguages waiting for your approval to merge

File: addons/remaining-replies/addon.json
Comment (WorldLanguages) [inline addons/remaining-replies/addon.json:31]:
```suggestion
  "versionAdded": "1.39.0",
```

**Comment by WorldLanguages** [comment]
Nice work
I'm not going to test it carefully again, it's a studio feature which is disabled by default.

**Pull request merged**
